### PR TITLE
Add Eth1 address withdrawal support

### LINKED ISSUE #188: Support ETH1_ADDRESS_WITHDRAWAL_PREFIX withdrawal credentials (0x01)
### Issue
Eth2 spec v1.0.1 enables a new withdrawal credential type that is a 20-byte eth1 address. (See https://github.com/ethereum/eth2.0-specs/pull/2149). Currently, `deposit-cli` doesn't support it.

### How to fix it
1. Enable user to type their Eth1 address
2. Check the given Eth1 address format
3. Set withdrawal credential with https://github.com/ethereum/eth2.0-specs/pull/2149 rule
4. Generate deposit data properly

### PR DESCRIPTION
**Pull request by hwwhww**
Fix #188, support https://github.com/ethereum/eth2.0-specs/pull/2149

- cli interface: add `--eth1_withdrawal_address` argument: If this field is set and valid, the given Eth1 address will be used to create the withdrawal credentials. Otherwise, it will generate withdrawal credentials with the mnemonic-derived withdrawal public key.
- `validate_deposit`: add public key and withdrawal credentials validations.

### CODE REVIEW

File: eth2deposit/cli/generate_keys.py [hunk 1]
```diff
@@ -57,6 +60,18 @@ def validate_password(cts: click.Context, param: Any, password: str) -> str:
     return password
 
 
+def validate_eth1_withdrawal_address(cts: click.Context, param: Any, address: str) -> HexAddress:
+    if address is None:
+        return None
+    if not is_hex_address(address):
```
Comment (hwwhww) [inline eth2deposit/cli/generate_keys.py:66]: Note: the prefix `0x` is optional here.

**Comment by hwwhww** [comment]
/cc @djrtwo 
Considering your de-Eth1/Eth2 rebranding, any plan to rename `ETH1_ADDRESS_WITHDRAWAL_PREFIX`?

File: eth2deposit/credentials.py [hunk 2]
```diff
@@ -57,10 +69,39 @@ def signing_pk(self) -> bytes:
...
     def withdrawal_credentials(self) -> bytes:
-        withdrawal_credentials = BLS_WITHDRAWAL_PREFIX
-        withdrawal_credentials += SHA256(self.withdrawal_pk)[1:]
+        if self.withdrawal_type == WithdrawalType.BLS_WITHDRAWAL:
+            withdrawal_credentials = BLS_WITHDRAWAL_PREFIX
+            withdrawal_credentials += SHA256(self.withdrawal_pk)[1:]
+        elif self.withdrawal_type == WithdrawalType.ETH1_ADDRESS_WITHDRAWAL:
+            withdrawal_credentials = ETH1_ADDRESS_WITHDRAWAL_PREFIX
+            withdrawal_credentials += b'\x00' * 11
+            withdrawal_credentials += self.eth1_withdrawal_address
```
Comment (CarlBeek) [inline eth2deposit/credentials.py:102]: Is it worth adding a sanity check here to handle the (hopefully impossible case) of when `self.eth1_withdrawal_address == None`?

File: eth2deposit/utils/validation.py [hunk 3]
```diff
@@ -45,6 +51,28 @@ def validate_deposit(deposit_data_dict: Dict[str, Any]) -> bool:
     deposit_message_root = bytes.fromhex(deposit_data_dict['deposit_data_root'])
     fork_version = bytes.fromhex(deposit_data_dict['fork_version'])
 
+    # Verify pubkey
+    if len(pubkey) != 48:
+        return False
+    if pubkey != credential.signing_pk:
+        return False
+
+    # Verify withdrawal credential
```
Comment (CarlBeek) [inline eth2deposit/utils/validation.py:60]: I appreciate the additional checks here. 👍

**Review by CarlBeek — CHANGES_REQUESTED** [review]
I like that this is being added. 👍 

One issue is that by adding this functionality to `generate_keys` it is automagically added to both the `new-mnemonic` and `existing-mnemonic` commands. The latter of which doesn't have tests and isn't mentioned in the `README`.

**Review by CarlBeek — APPROVED** [review]
LGTM! Let's get it merged.

File: eth2deposit/cli/generate_keys.py (hunk 1 above)
Comment (edmont87) [reply eth2deposit/cli/generate_keys.py:66]: hi

Comment (edmont87) [reply eth2deposit/cli/generate_keys.py:66]: hello word

**Review by G-11-P — COMMENTED** [review]
commit and merge

**Pull request merged**